pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/startree/AggregationFunctionColumn.java

```diff
@@ -127,9 +149,14 @@ public boolean equals(Object obj) {
     if (this == obj) {
       return true;
     }
-    if (obj instanceof AggregationFunctionColumnPair) {
-      AggregationFunctionColumnPair anotherPair = (AggregationFunctionColumnPair) obj;
-      return _functionType == anotherPair._functionType && _column.equals(anotherPair._column);
+    if (obj instanceof AggregationFunctionColumn) {
+      AggregationFunctionColumn other = (AggregationFunctionColumn) obj;
+      // TODO: Revisit this since it means that for aggregation functions where a certain config parameter need not be
+      // checked to determine whether a query can be served by a star-tree index, we won't rebuild a star-tree index
+      // if the parameter value is changed in the index configuration.
+      return _functionType == other._functionType && _column.equals(other._column)
+          && AggregationFunctionType.compareFunctionParametersForStarTree(_functionType, _functionParameters,
+          other._functionParameters) == 0;
```

Comment:
I believe we made it comparable because we are using it as the key in a `TreeMap` in `StarTreeV2Metadata`. Since currently we are adding a map inside, we can consider changing it to a hash map and do exact match.
We should also update the `hashCode()` method

Reply:
Actually after a second thought, there are usually not that many aggregations configured in the star-tree, especially for the same aggregate-column pair. We can consider managing the info as `Map<FunctionColumnPair, List<Pair<Map<String, Object>, AggregationSpec>>`.

Reply:
Hm actually I just realized that there's an assumption probably baked into many places that one star-tree can't have multiple pre-aggregations for the same aggregation function / column pair (see [here](https://github.com/apache/pinot/blob/d411f34b869aaecad803a1dde659a712c8e8c18a/pinot-core/src/main/java/org/apache/pinot/core/startree/plan/StarTreeProjectPlanNode.java#L61-L91) for instance where we're using the `functionName__columnName` to retrieve data from the star-tree). I think it might be cleaner overall to make this assumption explicit now that we're introducing configurable function parameters - essentially add a table config validation that prevents a single star-tree index configuration (`StarTreeIndexConfig`) from having multiple configurations for the same function / column in `StarTreeAggregationConfig`. So, if a user wants to have a pre-aggregation on `DISTINCTCOUNTHLL(col, 16)` as well as `DISTINCTCOUNTHLL(col, 8)` for some reason, these should be configured in separate star-trees. This way we can safely continue to use the existing logic for retrieving data from a star-tree index once it has already been matched for a query.

I agree with your comment on generalizing the logic of matching aggregations in `StarTreeUtils.createStarTreeBasedProjectOperator()` though. I like your idea of modifying the `AggregationFunction` interface to allow each aggregation to decide whether a particular star-tree index can be used. Although I'm thinking that given the above observations on not having multiple entities for the same function / column pair in a single star-tree, we can extract the function parameters out of this `AggregationFunctionColumn` class (reverting it to what it was earlier). We can instead maintain the function parameters alongside the existing info in something like a `Map<AggregationFunctionColumnPair, Pair<Map<String, Object>, AggregationSpec>` (which also seems easier to reason about) in `StarTreeV2BuilderConfig`, `StarTreeV2Metadata` etc. WDYT?

Reply:
Sounds good! This limitation comes with how we store columns in star-tree, and it is not easy to work around it without introducing new star-tree index format. I'm good enforcing it.
As for the map, should we consider adding the parameters into `AggregationSpec`?